File: ios/XiriGoEcommerce/Core/DesignSystem/Component/Skeleton.swift
import SwiftUI

// MARK: - SkeletonConstants

/// Internal constants for skeleton components.
/// Token source: `shared/design-tokens/components.json` (skeleton atom).
private enum SkeletonConstants {
    /// Default height for text-line skeleton placeholders (approximates `bodyMedium` line height).
    static let lineDefaultHeight: CGFloat = 14
}

// MARK: - SkeletonBox

/// Rectangular shimmer placeholder with configurable dimensions and corner radius.
///
/// Individual skeleton shapes are decorative and hidden from the accessibility tree.
/// Use the ``skeleton(visible:placeholder:)`` modifier to wrap skeleton layouts
/// with an accessible "Loading content" label.
///
/// Usage:
/// ```swift
/// SkeletonBox(width: 170, height: 170)
/// SkeletonBox(width: 170, height: 170, cornerRadius: XGCornerRadius.large)
/// ```
struct SkeletonBox: View {
    // MARK: - Lifecycle

    init(
        width: CGFloat,
        height: CGFloat,
        cornerRadius: CGFloat = XGCornerRadius.medium,
    ) {
        self.width = width
        self.height = height
        self.cornerRadius = cornerRadius
    }

    // MARK: - Internal

    let width: CGFloat
    let height: CGFloat
    let cornerRadius: CGFloat

    var body: some View {
        RoundedRectangle(cornerRadius: cornerRadius)
            .fill(XGColors.shimmer)
            .frame(width: width, height: height)
            .shimmerEffect()
            .accessibilityHidden(true)
    }
}

// MARK: - SkeletonLine

/// Text-line shimmer placeholder with a fixed small corner radius.
///
/// The default height of 14pt approximates `bodyMedium` line height from the typography scale,
/// making `SkeletonLine` immediately usable for most text placeholders without specifying height.
///
/// The corner radius is fixed at ``XGCornerRadius/small`` (6pt) and is not configurable,
/// ensuring consistent text-line appearance across the app.
///
/// Individual skeleton shapes are decorative and hidden from the accessibility tree.
struct SkeletonLine: View {
    // MARK: - Lifecycle

    init(
        width: CGFloat,
        height: CGFloat = SkeletonConstants.lineDefaultHeight,
    ) {
        self.width = width
        self.height = height
    }

    // MARK: - Internal

    let width: CGFloat
    let height: CGFloat

    var body: some View {
        RoundedRectangle(cornerRadius: XGCornerRadius.small)
            .fill(XGColors.shimmer)
            .frame(width: width, height: height)
            .shimmerEffect()
            .accessibilityHidden(true)
    }
}

// MARK: - SkeletonCircle

/// Circular shimmer placeholder for avatars, icons, and thumbnails.
///
/// Individual skeleton shapes are decorative and hidden from the accessibility tree.
struct SkeletonCircle: View {
    let size: CGFloat

    var body: some View {
        Circle()
            .fill(XGColors.shimmer)
            .frame(width: size, height: size)
            .shimmerEffect()
            .accessibilityHidden(true)
    }
}

// MARK: - SkeletonModifier

/// Content-wrapping modifier that crossfades between a skeleton placeholder and real content.
///
/// When `visible` is `true`, the placeholder is shown with an accessible "Loading content" label.
/// When `visible` transitions to `false`, a crossfade animation reveals the real content.
///
/// The crossfade duration is ``XGMotion/Crossfade/contentSwitch`` (0.2s).
struct SkeletonModifier<Placeholder: View>: ViewModifier {
    let visible: Bool
    let placeholder: Placeholder

    func body(content: Content) -> some View {
        Group {
            if visible {
                placeholder
                    .transition(.opacity)
                    .accessibilityLabel(Text("skeleton_loading_placeholder"))
            } else {
                content
                    .transition(.opacity)
            }
        }
        .animation(
            .easeInOut(duration: XGMotion.Crossfade.contentSwitch),
            value: visible,
        )
    }
}

// MARK: - View Extension

extension View {
    /// Wraps this view with a crossfade skeleton placeholder for loading states.
    ///
    /// When `visible` is `true`, the `placeholder` is displayed instead of the content.
    /// When `visible` transitions to `false`, the real content fades in using the
    /// ``XGMotion/Crossfade/contentSwitch`` duration.
    ///
    /// - Parameters:
    ///   - visible: When `true`, shows the placeholder. When `false`, shows content.
    ///   - placeholder: The skeleton layout to show during loading.
    /// - Returns: A view that crossfades between placeholder and content.
    func skeleton(
        visible: Bool,
        @ViewBuilder placeholder: () -> some View,
    ) -> some View {
        modifier(SkeletonModifier(visible: visible, placeholder: placeholder()))
    }
}

// MARK: - PreviewConstants

private enum PreviewConstants {
    static let boxWidth: CGFloat = 170
    static let boxHeight: CGFloat = 170
    static let lineWidthLong: CGFloat = 140
    static let lineWidthShort: CGFloat = 80
    static let lineSmallHeight: CGFloat = 12
    static let circleSize: CGFloat = 48
    static let ratingCircleSize: CGFloat = 12
    static let ratingLineWidth: CGFloat = 30
}

// MARK: - SkeletonBox Preview

#Preview("SkeletonBox") {
    VStack(spacing: XGSpacing.sm) {
        SkeletonBox(
            width: PreviewConstants.boxWidth,
            height: PreviewConstants.boxHeight,
        )
        SkeletonBox(
            width: PreviewConstants.boxWidth,
            height: PreviewConstants.boxHeight,
            cornerRadius: XGCornerRadius.large,
        )
    }
    .padding(XGSpacing.base)
    .xgTheme()
}

// MARK: - SkeletonLine Preview

#Preview("SkeletonLine") {
    VStack(alignment: .leading, spacing: XGSpacing.sm) {
        SkeletonLine(width: PreviewConstants.lineWidthLong)
        SkeletonLine(
            width: PreviewConstants.lineWidthShort,
            height: PreviewConstants.lineSmallHeight,
        )
    }
    .padding(XGSpacing.base)
    .xgTheme()
}

// MARK: - SkeletonCircle Preview

#Preview("SkeletonCircle") {
    VStack(spacing: XGSpacing.sm) {
        SkeletonCircle(size: PreviewConstants.circleSize)
        SkeletonCircle(size: PreviewConstants.ratingCircleSize)
    }
    .padding(XGSpacing.base)
    .xgTheme()
}

// MARK: - Product Card Skeleton Preview

#Preview("Product Card Skeleton") {
    VStack(alignment: .leading, spacing: XGSpacing.sm) {
        SkeletonBox(
            width: PreviewConstants.boxWidth,
            height: PreviewConstants.boxHeight,
        )
        SkeletonLine(width: PreviewConstants.lineWidthLong)
        SkeletonLine(
            width: PreviewConstants.lineWidthShort,
            height: PreviewConstants.lineSmallHeight,
        )
        HStack(spacing: XGSpacing.xs) {
            SkeletonCircle(size: PreviewConstants.ratingCircleSize)
            SkeletonLine(
                width: PreviewConstants.ratingLineWidth,
                height: PreviewConstants.lineSmallHeight,
            )
        }
    }
    .padding(XGSpacing.base)
    .xgTheme()
}

// MARK: - SkeletonCrossfadeDemo

/// Interactive demo showing the crossfade transition between skeleton placeholder and real content.
private struct SkeletonCrossfadeDemo: View {
    // MARK: - Internal

    var body: some View {
        VStack(alignment: .leading, spacing: XGSpacing.sm) {
            // Preview-only label (not user-facing)
            Text(verbatim: "Crossfade Demo (isLoading=\(isLoading))")
                .font(.caption)

            Text(verbatim: "Real content loaded!")
                .skeleton(visible: isLoading) {
                    VStack(alignment: .leading, spacing: XGSpacing.sm) {
                        SkeletonBox(
                            width: PreviewConstants.boxWidth,
                            height: PreviewConstants.boxHeight,
                        )
                        SkeletonLine(width: PreviewConstants.lineWidthLong)
                    }
                }

            Button {
                isLoading.toggle()
            } label: {
                // Preview-only toggle label (not user-facing)
                Text(verbatim: isLoading ? "Show Content" : "Show Skeleton")
            }
        }
        .padding(XGSpacing.base)
    }

    // MARK: - Private

    @State private var isLoading = true
}

#Preview("Skeleton Crossfade") {
    SkeletonCrossfadeDemo()
        .xgTheme()
}

File: android/app/src/main/java/com/xirigo/ecommerce/core/designsystem/component/Skeleton.kt
package com.xirigo.ecommerce.core.designsystem.component

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.xirigo.ecommerce.R
import com.xirigo.ecommerce.core.designsystem.theme.XGColors
import com.xirigo.ecommerce.core.designsystem.theme.XGCornerRadius
import com.xirigo.ecommerce.core.designsystem.theme.XGMotion
import com.xirigo.ecommerce.core.designsystem.theme.XGSpacing
import com.xirigo.ecommerce.core.designsystem.theme.XGTheme

// Token source: components/atoms/xg-skeleton.json

private val SkeletonLineDefaultHeight = 14.dp

// Preview constants
private val PreviewBoxWidth = 170.dp
private val PreviewBoxHeight = 170.dp
private val PreviewLineWidthLong = 140.dp
private val PreviewLineWidthShort = 80.dp
private val PreviewLineSmallHeight = 12.dp
private val PreviewCircleSize = 48.dp
private val PreviewRatingCircleSize = 12.dp
private val PreviewRatingLineWidth = 30.dp

/**
 * Rectangular shimmer placeholder with configurable dimensions and corner radius.
 *
 * Individual skeleton shapes are decorative and hidden from the accessibility tree.
 * Use [XGSkeleton] to wrap skeleton layouts with an accessible "Loading content" label.
 */
@Composable
fun SkeletonBox(
    width: Dp,
    height: Dp,
    modifier: Modifier = Modifier,
    cornerRadius: Dp = XGCornerRadius.Medium,
) {
    val shape = RoundedCornerShape(cornerRadius)
    Box(
        modifier = modifier
            .size(width, height)
            .clip(shape)
            .background(XGColors.Shimmer)
            .shimmerEffect(),
    )
}

/**
 * Text-line shimmer placeholder with a fixed small corner radius.
 *
 * The default height of 14dp approximates `bodyMedium` line height from the typography scale,
 * making [SkeletonLine] immediately usable for most text placeholders without specifying height.
 *
 * Individual skeleton shapes are decorative and hidden from the accessibility tree.
 */
@Composable
fun SkeletonLine(
    width: Dp,
    modifier: Modifier = Modifier,
    height: Dp = SkeletonLineDefaultHeight,
) {
    val shape = RoundedCornerShape(XGCornerRadius.Small)
    Box(
        modifier = modifier
            .size(width, height)
            .clip(shape)
            .background(XGColors.Shimmer)
            .shimmerEffect(),
    )
}

/**
 * Circular shimmer placeholder for avatars, icons, and thumbnails.
 *
 * Individual skeleton shapes are decorative and hidden from the accessibility tree.
 */
@Composable
fun SkeletonCircle(size: Dp, modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .size(size)
            .clip(CircleShape)
            .background(XGColors.Shimmer)
            .shimmerEffect(),
    )
}

/**
 * Content-wrapping composable that crossfades between a skeleton [placeholder] and real [content].
 *
 * When [visible] is `true`, the [placeholder] is shown with an accessible "Loading content" label.
 * When [visible] transitions to `false`, a crossfade animation reveals the real [content].
 *
 * The crossfade duration is [XGMotion.Crossfade.CONTENT_SWITCH] (200ms).
 *
 * @param visible When `true`, shows the placeholder. When `false`, shows content.
 * @param placeholder The skeleton layout to show during loading.
 * @param modifier Modifier applied to the [AnimatedContent] container.
 * @param content The real content to show when loading completes.
 */
@Composable
fun XGSkeleton(
    visible: Boolean,
    placeholder: @Composable () -> Unit,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    val loadingDescription = stringResource(R.string.skeleton_loading_placeholder)

    AnimatedContent(
        targetState = visible,
        modifier = modifier.semantics {
            if (visible) {
                contentDescription = loadingDescription
            }
        },
        transitionSpec = {
            fadeIn(tween(XGMotion.Crossfade.CONTENT_SWITCH)) togetherWith
                fadeOut(tween(XGMotion.Crossfade.CONTENT_SWITCH))
        },
        label = "skeletonCrossfade",
    ) { isLoading ->
        if (isLoading) placeholder() else content()
    }
}

// region Preview

@Preview(showBackground = true)
@Composable
private fun SkeletonBoxPreview() {
    XGTheme {
        Column(
            modifier = Modifier.padding(XGSpacing.Base),
            verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
        ) {
            SkeletonBox(width = PreviewBoxWidth, height = PreviewBoxHeight)
            SkeletonBox(
                width = PreviewBoxWidth,
                height = PreviewBoxHeight,
                cornerRadius = XGCornerRadius.Large,
            )
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun SkeletonLinePreview() {
    XGTheme {
        Column(
            modifier = Modifier.padding(XGSpacing.Base),
            verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
        ) {
            SkeletonLine(width = PreviewLineWidthLong)
            SkeletonLine(width = PreviewLineWidthShort, height = PreviewLineSmallHeight)
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun SkeletonCirclePreview() {
    XGTheme {
        Column(
            modifier = Modifier.padding(XGSpacing.Base),
            verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
        ) {
            SkeletonCircle(size = PreviewCircleSize)
            SkeletonCircle(size = PreviewRatingCircleSize)
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun ProductCardSkeletonPreview() {
    XGTheme {
        Column(
            modifier = Modifier.padding(XGSpacing.Base),
            verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
        ) {
            SkeletonBox(width = PreviewBoxWidth, height = PreviewBoxHeight)
            SkeletonLine(width = PreviewLineWidthLong)
            SkeletonLine(width = PreviewLineWidthShort, height = PreviewLineSmallHeight)
            Row(horizontalArrangement = Arrangement.spacedBy(XGSpacing.XS)) {
                SkeletonCircle(size = PreviewRatingCircleSize)
                SkeletonLine(width = PreviewRatingLineWidth, height = PreviewLineSmallHeight)
            }
        }
    }
}

@Preview(showBackground = true)
@Composable
private fun XGSkeletonCrossfadePreview() {
    XGTheme {
        var isLoading by remember { mutableStateOf(true) }

        Column(
            modifier = Modifier.padding(XGSpacing.Base),
            verticalArrangement = Arrangement.spacedBy(XGSpacing.SM),
        ) {
            Text(
                text = "Crossfade Demo (isLoading=$isLoading)",
                style = MaterialTheme.typography.labelMedium,
            )

            XGSkeleton(
                visible = isLoading,
                placeholder = {
                    Column(verticalArrangement = Arrangement.spacedBy(XGSpacing.SM)) {
                        SkeletonBox(width = PreviewBoxWidth, height = PreviewBoxHeight)
                        SkeletonLine(width = PreviewLineWidthLong)
                    }
                },
            ) {
                Text(
                    text = "Real content loaded!",
                    style = MaterialTheme.typography.bodyLarge,
                )
            }

            Spacer(modifier = Modifier.height(XGSpacing.SM))

            // Toggle button for interactive preview testing
            androidx.compose.material3.Button(onClick = { isLoading = !isLoading }) {
                Text(text = if (isLoading) "Show Content" else "Show Skeleton")
            }
        }
    }
}

// endregion
